### PR TITLE
Bug 1694173: Improve status reporting

### PR DESCRIPTION
**Pull request by sallyom**
Currently we always set Available=False when we set Progressing=True and Available=True/Progressing=False.  This PR fine-tunes those to set independently when appropriate.

### CODE REVIEW

File: pkg/operator2/operator.go [hunk 1]
```diff
@@ -407,160 +412,175 @@ func (c *authOperator) checkReady(
...
 
-	wellknownReady, wellknownMsg, err := c.checkWellknownEndpointReady(authConfig, route)
+	wellknownReady, err := c.checkWellknownEndpointReady(authConfig, route, operatorConfig)
 	if err != nil {
 		return wellknownReady, fmt.Errorf("unable to check the .well-known endpoint: %v", err)
 	}
 	if !wellknownReady {
-		c.setProgressingStatus(operatorConfig, "WellknownNotReady", wellknownMsg)
+		setProgressingTrueAndAvailableFalse(operatorConfig, "WellKnownNotReady", err.Error())
 		return wellknownReady, nil
```
Comment (sallyom) [inline pkg/operator2/operator.go]: need to fix that ^^

File: pkg/operator2/operator.go [hunk 2]
```diff
@@ -407,160 +412,175 @@ func (c *authOperator) checkReady(
...
 	}
 
-	routeReady, routeMsg, err := c.checkRouteHealthy(route, routerSecret)
+	routeReady, err := c.checkRouteHealthy(route, routerSecret, operatorConfig)
 	if err != nil {
 		return routeReady, fmt.Errorf("unable to check route health: %v", err)
 	}
 	if !routeReady {
-		c.setProgressingStatus(operatorConfig, "RouteNotReady", routeMsg)
+		setProgressingTrueAndAvailableFalse(operatorConfig, "RouteNotReady", err.Error())
```
Comment (sallyom) [inline pkg/operator2/operator.go]: ^^ need to fix that

File: pkg/operator2/operator.go [hunk 3]
```diff
@@ -376,7 +380,8 @@ func (c *authOperator) handleSync(operatorConfig *operatorv1.Authentication) err
 		if c.versionGetter.GetVersions()[operatorSelfName] != operatorVersion {
 			c.versionGetter.SetVersion(operatorSelfName, operatorVersion)
 		}
-		c.setAvailableStatus(operatorConfig)
+		setAvailableTrue(operatorConfig, "ManagedResourcesSuccessfullyApplied")
```
Comment (sallyom) [inline pkg/operator2/operator.go]: not sure about that msg

File: pkg/operator2/operator.go [hunk 4]
```diff
@@ -407,160 +412,175 @@ func (c *authOperator) checkReady(
...
 	}
 
-	oauthClientsReady, oauthClientsMsg, err := c.oauthClientsReady(route)
+	oauthClientsReady, err := c.oauthClientsReady(route, operatorConfig)
 	if err != nil {
 		return oauthClientsReady, fmt.Errorf("unable to check OAuth clients' readiness: %v", err)
 	}
 	if !oauthClientsReady {
-		c.setProgressingStatus(operatorConfig, "OAuthClientsNotReady", oauthClientsMsg)
+		setProgressingTrueAndAvailableFalse(operatorConfig, "OAuthClientNotReady", err.Error())
```
Comment (sallyom) [inline pkg/operator2/operator.go]: fix that, too

**Comment by sallyom** [comment]
@enj I tested an upgrade to this PR payload and guess what? authentication CO was _not_ the very last CO to report new version & Available=True, first time I've seen that.  Upgrade was successful.

File: pkg/operator2/operator.go [hunk 5]
```diff
@@ -430,40 +435,46 @@ func (c *authOperator) checkReady(
...
 
 	if osinDeployment.DeletionTimestamp != nil {
-		return false, "", fmt.Errorf("the deployment is being deleted")
+		setProgressingTrueAndAvailableFalse(operatorConfig, reason, "deployment is being deleted")
+		return false, nil
 	}
 
 	if osinDeployment.Status.UpdatedReplicas != osinDeployment.Status.Replicas || osinDeployment.Status.UnavailableReplicas > 0 {
-		return false, "not all deployment replicas are ready", nil
+		setProgressingTrue(operatorConfig, reason, "not all deployment replicas are ready")
```
Comment (enj) [inline pkg/operator2/operator.go]: We are available here as long as we have one ready replica.

File: pkg/operator2/operator.go (hunk 3 above)
Comment (enj) [inline pkg/operator2/operator.go]: `AsExpected`?

Comment (sallyom) [reply pkg/operator2/operator.go]: that works

File: pkg/operator2/operator.go (hunk 5 above)
Comment (sallyom) [reply pkg/operator2/operator.go]: right, updated, missed that!

**Pull request merged**